unit_tests/log/test_main_log_recovery_parallel.cpp

```diff
@@ -64,43 +67,104 @@ void execute_test (const log_recovery_test_config &a_test_config,
 		<< "\t pages=" << a_database_config.max_page_count_per_volume
 		<< "\t parallel=" << a_test_config.parallel_count
 		<< "\t jobs=" << a_test_config.redo_job_count
+		<< "\t wait_past_previous_log_lsa=" << std::boolalpha << a_test_config.wait_past_previous_log_lsa
 		<< std::endl;
     }
 
+  // simulated database with volumes and pages
+  //
   ux_ut_database db_online { new ut_database (a_database_config) };
   ux_ut_database db_recovery { new ut_database (a_database_config) };
 
-  cublog::minimum_log_lsa_monitor minimum_log_lsa;
   log_rv_redo_context dummy_redo_context (NULL_LSA, log_reader::fetch_mode::NORMAL);
 
-  cublog::redo_parallel log_redo_parallel (a_test_config.parallel_count, &minimum_log_lsa, dummy_redo_context);
+  cublog::redo_parallel log_redo_parallel (a_test_config.parallel_count, true, dummy_redo_context);
 
+  // infrastructure to check progress of the test
+  //
+  struct
+  {
+    log_lsa m_log_lsa { MAX_LSA }; // NULL_LSA is the stop condition
+    std::mutex m_mtx;
+    std::condition_variable m_cv;
+  } wait_past_log_lsa_info;
+  std::thread wait_past_log_lsa_thr;
+  if (a_test_config.wait_past_previous_log_lsa)
+    {
+      wait_past_log_lsa_thr = std::thread ([&log_redo_parallel, &wait_past_log_lsa_info] ()
+      {
+	log_lsa local_log_lsa { MAX_LSA };
+	do
+	  {
+	    {
+	      std::unique_lock<std::mutex> ulock { wait_past_log_lsa_info.m_mtx };
+	      wait_past_log_lsa_info.m_cv.wait_for (ulock, std::chrono::milliseconds (100));
+	      local_log_lsa = wait_past_log_lsa_info.m_log_lsa;
+	    }
+	    if (!local_log_lsa.is_max () && !local_log_lsa.is_null ())
+	      {
+		log_redo_parallel.wait_past_target_lsa (local_log_lsa);
+	      }
+	  }
+	while (!local_log_lsa.is_null ());
+      });
+    }
+
+  REQUIRE (log_redo_parallel.get_calculated_minimum_not_applied_log_lsa ().is_max ());
+
+  // the test proper
+  //
   ut_database_values_generator global_values{ a_database_config };
   for (size_t idx = 0u; idx < a_test_config.redo_job_count; ++idx)
     {
       ux_ut_redo_job_impl job = db_online->generate_changes (*db_recovery, global_values);
 
       if (job->is_volume_creation () || job->is_page_creation ())
 	{
+	  if (a_test_config.wait_past_previous_log_lsa)
+	    {
+	      log_redo_parallel.set_outer_not_applied_log_lsa (job->get_log_lsa ());
```

Comment:
Normally `set_outer_not_applied_log_lsa()` should be called for each log record, regardless if it is applied by the main therad or it will be applied by a worker thread.